llvm/lib/Object/BuildID.cpp

```diff
@@ -24,20 +24,27 @@ using namespace llvm::object;
 namespace {
 
 template <typename ELFT> BuildIDRef getBuildID(const ELFFile<ELFT> &Obj) {
+  auto findBuildID = [&Obj](const auto &ShdrOrPhdr,
+                            uint64_t Alignment) -> std::optional<BuildIDRef> {
+    Error Err = Error::success();
+    for (auto N : Obj.notes(ShdrOrPhdr, Err))
+      if (N.getType() == ELF::NT_GNU_BUILD_ID &&
+          N.getName() == ELF::ELF_NOTE_GNU)
+        return N.getDesc(Alignment);
+    consumeError(std::move(Err));
+    return std::nullopt;
+  };
+
   auto Sections = cantFail(Obj.sections());
   if (!Sections.empty()) {
     for (const auto &S : Sections) {
       if (S.sh_type != ELF::SHT_NOTE)
         continue;
-      Error Err = Error::success();
-      for (auto N : Obj.notes(S, Err))
-        if (N.getType() == ELF::NT_GNU_BUILD_ID &&
-            N.getName() == ELF::ELF_NOTE_GNU)
-          return N.getDesc(S.sh_addralign);
-      consumeError(std::move(Err));
+      auto ShdrRes = findBuildID(S, S.sh_addralign);
+      if (ShdrRes)
+        return ShdrRes.value();
```

Comment:
This can be shortened to the suggestion.

Please also review the documentation on `auto` in the coding standards, as I think you're over-using it.
```suggestion
      if (std::optional<BuildIDRef> ShdrRes = findBuildID(S, S.sh_addralign))
        return ShdrRes.value();
```
(and the same style below for the Phdrs).

Reply:
Thanks for your suggestion, I will refer this code.
I'm sorry for over-using `auto`, I should read the code standard before I submit my code.
I'm really sorry about this.